### ISSUE TITLE
grid_sample Segmentation fault on CPU

### ISSUE THREAD

**Issue by zekunhao1995** [pytorch/pytorch#19826]
## 🐛 Bug
CPU version of grid_sample got segmentation fault under some 

circumstances

## To Reproduce
Run this:
import torch
import torch.nn.functional as F

img = torch.empty([1, 3, 123, 456], dtype=torch.float32)
offset = torch.zeros([1, 123, 456, 2], dtype=torch.float32, requires_grad=True)
optimizer = torch.optim.SGD([offset], lr=0.0005, momentum=0.9)

for i in range(200):
    optimizer.zero_grad()
    warped = F.grid_sample(img, offset, mode='bilinear',padding_mode="border")
    loss = torch.mean(torch.sqrt(torch.abs(offset)))
    loss.backward()
    optimizer.step()
    print('{:04d} {:.6f}, '.format(i, loss.item()))

## Environment
PyTorch version: 1.0.0
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 18.04.2 LTS
GCC version: (Ubuntu 7.3.0-27ubuntu1~18.04) 7.3.0
CMake version: version 3.10.2

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce GTX 1080 Ti
Nvidia driver version: 410.93
cuDNN version: Could not collect

Versions of relevant libraries:
[pip3] numpy==1.16.1
[pip3] torch==1.0.0
[pip3] torchfile==0.1.0
[pip3] torchvision==0.2.1
[pip3] torchviz==0.0.1
[conda] Could not collect

## Additional context
GDB backtrace:
(gdb) run
Starting program: /usr/bin/python3.6 try.py
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7ffff3aa5700 (LWP 18385)]
[New Thread 0x7ffff32a4700 (LWP 18386)]
[New Thread 0x7fffeeaa3700 (LWP 18387)]
[New Thread 0x7fffec2a2700 (LWP 18388)]
[New Thread 0x7fffe9aa1700 (LWP 18389)]
[New Thread 0x7fffe72a0700 (LWP 18390)]
[New Thread 0x7fffe4a9f700 (LWP 18391)]
[Thread 0x7fffe4a9f700 (LWP 18391) exited]
[Thread 0x7fffe72a0700 (LWP 18390) exited]
[Thread 0x7fffe9aa1700 (LWP 18389) exited]
[Thread 0x7fffec2a2700 (LWP 18388) exited]
[Thread 0x7fffeeaa3700 (LWP 18387) exited]
[Thread 0x7ffff32a4700 (LWP 18386) exited]
[Thread 0x7ffff3aa5700 (LWP 18385) exited]
[New Thread 0x7fffe4a9f700 (LWP 18394)]
[New Thread 0x7fffe72a0700 (LWP 18395)]
[New Thread 0x7fffe9aa1700 (LWP 18396)]
[New Thread 0x7fffec2a2700 (LWP 18398)]
[New Thread 0x7fff7a11c700 (LWP 18399)]
[New Thread 0x7fff7991b700 (LWP 18400)]
0000 0.010665, 

Thread 1 "python3.6" received signal SIGSEGV, Segmentation fault.
0x00007fff9823141f in std::enable_if<((((4l)==(1))||((4l)==(2)))||((4l)==(4)))||((4l)==(8)), at::vec256::(anonymous namespace)::Vec256<float> >::type at::vec256::(anonymous namespace)::mask_gather<4l, float>(at::vec256::(anonymous namespace)::Vec256<float> const&, float const*, at::vec256::(anonymous namespace)::Vec256<at::vec256::(anonymous namespace)::int_of_size<sizeof (float)>::type> const&, at::vec256::(anonymous namespace)::Vec256<float>&) [clone .constprop.166] ()
   from /usr/local/lib/python3.6/dist-packages/torch/lib/libcaffe2.so
(gdb) backtrace
#0  0x00007fff9823141f in std::enable_if<((((4l)==(1))||((4l)==(2)))||((4l)==(4)))||((4l)==(8)), at::vec256::(anonymous namespace)::Vec256<float> >::type at::vec256::(anonymous namespace)::mask_gather<4l, float>(at::vec256::(anonymous namespace)::Vec256<float> const&, float const*, at::vec256::(anonymous namespace)::Vec256<at::vec256::(anonymous namespace)::int_of_size<sizeof (float)>::type> const&, at::vec256::(anonymous namespace)::Vec256<float>&) [clone .constprop.166] ()
   from /usr/local/lib/python3.6/dist-packages/torch/lib/libcaffe2.so
#1  0x00007fff9824d7a1 in at::native::(anonymous namespace)::grid_sampler_2d_cpu_kernel_impl(at::Tensor const&, at::Tensor const&, long, long)::{lambda()#1}::operator()() const::{lambda()#2}::operator()() const ()
   from /usr/local/lib/python3.6/dist-packages/torch/lib/libcaffe2.so
#2  0x00007fff9824ee10 in at::native::(anonymous namespace)::grid_sampler_2d_cpu_kernel_impl(at::Tensor const&, at::Tensor const&, long, long) ()
   from /usr/local/lib/python3.6/dist-packages/torch/lib/libcaffe2.so
#3  0x00007fff979d5c3d in at::native::grid_sampler_2d_cpu(at::Tensor const&, at::Tensor const&, long, long) ()
   from /usr/local/lib/python3.6/dist-packages/torch/lib/libcaffe2.so
#4  0x00007fff97b97954 in at::CPUFloatType::grid_sampler_2d(at::Tensor const&, at::Tensor const&, long, long) const ()
   from /usr/local/lib/python3.6/dist-packages/torch/lib/libcaffe2.so
#5  0x00007fff95b3b8cb in torch::autograd::VariableType::grid_sampler_2d(at::Tensor const&, at::Tensor const&, long, long) const ()
   from /usr/local/lib/python3.6/dist-packages/torch/lib/libtorch.so.1
---Type <return> to continue, or q <return> to quit---
#6  0x00007fff979d6780 in at::native::grid_sampler(at::Tensor const&, at::Tensor const&, long, long) () from /usr/local/lib/python3.6/dist-packages/torch/lib/libcaffe2.so
#7  0x00007fff97c4f344 in at::TypeDefault::grid_sampler(at::Tensor const&, at::Tensor const&, long, long) const () from /usr/local/lib/python3.6/dist-packages/torch/lib/libcaffe2.so
#8  0x00007fff95b550e3 in torch::autograd::VariableType::grid_sampler(at::Tensor const&, at::Tensor const&, long, long) const () from /usr/local/lib/python3.6/dist-packages/torch/lib/libtorch.so.1
#9  0x00007ffff2e1ae78 in torch::autograd::THPVariable_grid_sampler(_object*, _object*, _object*) () from /usr/local/lib/python3.6/dist-packages/torch/lib/libtorch_python.so
#10 0x00000000005030d5 in ?? ()
#11 0x0000000000506859 in _PyEval_EvalFrameDefault ()
#12 0x0000000000504c28 in ?? ()
#13 0x0000000000502540 in ?? ()
#14 0x0000000000502f3d in ?? ()
#15 0x0000000000507641 in _PyEval_EvalFrameDefault ()
#16 0x0000000000501945 in _PyFunction_FastCallDict ()
#17 0x0000000000591461 in ?? ()
#18 0x000000000059ebbe in PyObject_Call ()
#19 0x0000000000507c17 in _PyEval_EvalFrameDefault ()
#20 0x0000000000504c28 in ?? ()
#21 0x0000000000501b2e in _PyFunction_FastCallDict ()
#22 0x0000000000591461 in ?? ()
#23 0x000000000059ebbe in PyObject_Call ()
#24 0x000000000054d4e2 in ?? ()
#25 0x00000000005a730c in _PyObject_FastCallKeywords ()
#26 0x0000000000503073 in ?? ()
#27 0x0000000000506859 in _PyEval_EvalFrameDefault ()
#28 0x0000000000504c28 in ?? ()
#29 0x0000000000506393 in PyEval_EvalCode ()
#30 0x0000000000634d52 in ?? ()
#31 0x0000000000634e0a in PyRun_FileExFlags ()
#32 0x00000000006385c8 in PyRun_SimpleFileExFlags ()
#33 0x000000000063915a in Py_Main ()
#34 0x00000000004a6f10 in main ()

**Comment by SsnL** [comment]
it segfaults when the grid has nan. we should fix this.

**Comment by SsnL** [comment]
`torch.mean(torch.sqrt(torch.abs(offset)))` this gives you nan grad since your init values are zeros.

**Comment by zekunhao1995** [comment]
Hi Tongzhou, thanks for the prompt reply! That looks exactly like where the issue is.

**Comment by SsnL** [comment]
yeah, https://github.com/pytorch/pytorch/blob/1071e92335d8a69107f4694f56bebbb9655371aa/aten/src/ATen/cpu/vec256/vec256.h#L115-L125
can't handle nan...

**Comment by SsnL** [comment]
Simplest "fix" to add range checking to that function using signaling `_mm256_cmp_ps`. A more performant way is to compare as ints.
